File: widgetssdk/src/main/java/com/glia/widgets/filepreview/ui/FilePreviewController.java
package com.glia.widgets.filepreview.ui;

import com.glia.widgets.core.engagement.domain.GliaOnEngagementEndUseCase;
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromCacheUseCase;
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromDownloadsUseCase;
import com.glia.widgets.filepreview.domain.usecase.PutImageFileToDownloadsUseCase;
import com.glia.widgets.helper.Logger;

import io.reactivex.disposables.Disposable;

public class FilePreviewController
        implements FilePreviewContract.Controller, GliaOnEngagementEndUseCase.Listener {
    private static final String TAG = FilePreviewController.class.getSimpleName();

    private FilePreviewContract.View view;
    private State state = new State();

    private final GetImageFileFromDownloadsUseCase getImageFileFromDownloadsUseCase;
    private final GetImageFileFromCacheUseCase getImageFileFromCacheUseCase;
    private final PutImageFileToDownloadsUseCase putImageFileToDownloadsUseCase;
    private final GliaOnEngagementEndUseCase onEngagementEndUseCase;

    private Disposable disposable;

    public FilePreviewController(
            GetImageFileFromDownloadsUseCase getImageFileFromDownloadsUseCase,
            GetImageFileFromCacheUseCase getImageFileFromCacheUseCase,
            PutImageFileToDownloadsUseCase putImageFileToDownloadsUseCase,
            GliaOnEngagementEndUseCase onEngagementEndUseCase
    ) {
        this.getImageFileFromDownloadsUseCase = getImageFileFromDownloadsUseCase;
        this.getImageFileFromCacheUseCase = getImageFileFromCacheUseCase;
        this.putImageFileToDownloadsUseCase = putImageFileToDownloadsUseCase;
        this.onEngagementEndUseCase = onEngagementEndUseCase;
    }

    public void setView(FilePreviewContract.View view) {
        this.view = view;
        this.state.reset();
        onEngagementEndUseCase.execute(this);
    }

    @Override
    public void onImageDataReceived(String bitmapId, String bitmapName) {
        setState(state.setImageData(bitmapId, bitmapName));
    }

    @Override
    public void onImageRequested() {
        setState(state.setImageLoadingFromDownloads());
        disposable = getImageFileFromDownloadsUseCase.execute(state.getImageIdName())
                .subscribe(
                        image -> setState(
                                state.setImageLoadedFromDownloads()
                                        .setLoadedImage(image)
                        ),
                        error -> onRequestImageFromCache()
                );
    }

    @Override
    public void onSharePressed() {
        view.shareImageFile(state.getImageIdName());
    }

    @Override
    public void onDownloadPressed() {
        disposable = putImageFileToDownloadsUseCase
                .execute(state.getImageIdName(), state.getLoadedImage())
                .subscribe(
                        () -> {
                            setState(state.setImageLoadedFromDownloads());
                            view.showOnImageSaveSuccess();
                        },
                        error -> view.showOnImageSaveFailed()
                );
    }

    @Override
    public void onDestroy() {
        if (disposable != null) disposable.dispose();
        onEngagementEndUseCase.unregisterListener(this);
        state.reset();
    }

    @Override
    public void engagementEnded() {
        Logger.d(TAG, "engagementEndedByOperator");
        view.engagementEnded();
    }

    private synchronized void setState(State state) {
        this.state = state;
        if (view != null) view.onStateUpdated(state);
    }

    private void onRequestImageFromCache() {
        setState(state.setImageLoadingFromCache());
        disposable = getImageFileFromCacheUseCase.execute(state.getImageIdName())
                .subscribe(
                        image -> setState(state.setImageLoadedFromCache().setLoadedImage(image)),
                        error -> {
                            view.showOnImageLoadingFailed();
                            setState(state.setImageLoadingFailure());
                        }
                );
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/filepreview/ui/FilePreviewController.kt
package com.glia.widgets.filepreview.ui

import android.net.Uri
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromCacheUseCase
import com.glia.widgets.filepreview.domain.usecase.GetImageFileFromDownloadsUseCase
import com.glia.widgets.filepreview.domain.usecase.PutImageFileToDownloadsUseCase
import io.reactivex.rxjava3.disposables.CompositeDisposable

internal class FilePreviewController @JvmOverloads constructor(
    private val getImageFileFromDownloadsUseCase: GetImageFileFromDownloadsUseCase,
    private val getImageFileFromCacheUseCase: GetImageFileFromCacheUseCase,
    private val putImageFileToDownloadsUseCase: PutImageFileToDownloadsUseCase,
    private var state: State = State(),
    private val disposables: CompositeDisposable = CompositeDisposable()
) : FilePreviewContract.Controller {
    private var view: FilePreviewContract.View? = null

    override fun setView(view: FilePreviewContract.View) {
        this.view = view
        state = State()
    }

    @Synchronized
    private fun setState(state: State) {
        this.state = state
        view?.onStateUpdated(state)
    }

    override fun onImageDataReceived(bitmapId: String, bitmapName: String) {
        setState(state.withImageData(bitmapId, bitmapName))
    }

    override fun onImageRequested() {
        setState(state.loadingImageFromDownloads())

        disposables.add(
            getImageFileFromDownloadsUseCase(state.imageIdName)
                .subscribe(
                    { setState(state.withImageLoadedFromDownloads(it)) }
                ) { onRequestImageFromCache() }
        )
    }

    override fun onSharePressed() {
        if (state.imageLoadingState == State.ImageLoadingState.LOCAL) {
            view?.shareImageFile(state.localImageUri ?: return)
        } else {
            view?.shareImageFile(state.imageIdName)
        }
    }

    override fun onDownloadPressed() {
        disposables.add(
            putImageFileToDownloadsUseCase(state.imageIdName, state.loadedImage ?: return)
                .subscribe(
                    {
                        setState(state.withImageLoadedFromDownloads())
                        view?.showOnImageSaveSuccess()
                    }
                ) {
                    view?.showOnImageSaveFailed()
                }
        )
    }

    private fun onRequestImageFromCache() {
        setState(state.loadingImageFromCache())

        disposables.add(
            getImageFileFromCacheUseCase(state.imageIdName)
                .subscribe({
                    setState(state.withImageLoadedFromCache(it))
                }) {
                    view?.showOnImageLoadingFailed()
                    setState(state.imageLoadingFiled())
                }
        )
    }

    override fun onDestroy() {
        disposables.clear()
        state = State()
    }

    override fun onLocalImageReceived(uri: Uri) {
        setState(state.withLocalImage(uri))
    }
}
